emerging_optimizers/orthogonalized_optimizers/adaptive_muon.py

```diff
@@ -0,0 +1,242 @@
+                moment2 = torch.zeros(moment2_shape, dtype=grad.dtype, device=grad.device)
+            else:
+                raise TypeError(f"Invalid second moment method: {self.moment2_method}")
+
+            state["moment2_buffer"] = moment2
+
+    def _apply_moment2_normalization(
+        self,
+        orth_grad: torch.Tensor,
+        moment2: torch.Tensor,
+        beta2: float,
+        eps: float,
+    ) -> torch.Tensor:
+        """Apply AdamW-style second moment accumulation and normalization.
+
+        This method supports two variants:
+        - "adamuon": Full elementwise second moment (like AdamW, https://arxiv.org/abs/2507.11005)
+        - "normuon": Row or column-wise second moment (https://arxiv.org/abs/2510.05491)
+
+        For both methods:
+        1. Updates the second moment as an EMA of squared gradients
+        2. Returns the adaptively scaled gradient
+
+        Args:
+            orth_grad: The orthogonalized gradient tensor.
+            moment2: The second moment buffer from state.
+            beta2: The exponential decay rate for second moment.
+            eps: Small constant for numerical stability.
+
+        Returns:
+            The adaptively scaled weight update tensor.
+        """
+        if self.moment2_method == "adamuon":
+            # AdamMuon: Full elementwise second moment like AdamW
+            # Update second moment with EMA of squared orthogonalized gradient
+            moment2.lerp_(orth_grad.square(), 1 - beta2)
+
+            # AdamW-style division: grad / (sqrt(moment2) + eps)
+            denom = moment2.sqrt() + eps
+            return orth_grad / denom
+
+        elif self.moment2_method == "normuon":
+            # NorMuon: Row or column-wise second moment
+            # Compute mean of squared gradients along one dimension based on shape
+            # Average along the longer dimension to preserve structure along shorter dim
+            avg_dim = -1 if orth_grad.shape[-2] >= orth_grad.shape[-1] else -2
+            v_mean = orth_grad.square().mean(dim=avg_dim, keepdim=True)
+
+            # Update second moment with EMA
+            moment2.lerp_(v_mean, 1 - beta2)
+
+            # NorMuon uses reciprocal square root with clamping
+            step_size = moment2.clamp_min(eps).rsqrt_()
+            return orth_grad * step_size
+
+        else:
+            raise TypeError(f"Invalid second moment method: {self.moment2_method}")
+
+    @torch.no_grad()  # type: ignore[misc]
+    @override
+    def step(self, closure: Callable[[], float] | None = None) -> float | None:
+        """Performs a single optimization step with second moment normalization.
+
+        Args:
+            closure: A closure that reevaluates the model and returns the loss.
+        """
+        if closure is None:
+            loss = None
+        else:
+            loss = closure()
+
+        for group in self.param_groups:
+            for p in group["params"]:
+                if p.dim() == 1:
```

Comment:
I know we did this in base class, but I think we should check dim == 2. nd is also not supported.